src/modules/powerrename/lib/PowerRenameManager.cpp

```diff
@@ -0,0 +1,1000 @@
+#include "stdafx.h"
+#include "PowerRenameManager.h"
+#include "PowerRenameRegEx.h" // Default RegEx handler
+#include <algorithm>
+#include <shlobj.h>
+#include "helpers.h"
+#include <filesystem>
+namespace fs = std::filesystem;
+
+extern HINSTANCE g_hInst;
+
+// The default FOF flags to use in the rename operations
+#define FOF_DEFAULTFLAGS (FOF_ALLOWUNDO | FOFX_ADDUNDORECORD | FOFX_SHOWELEVATIONPROMPT | FOF_RENAMEONCOLLISION)
+
+IFACEMETHODIMP_(ULONG) CPowerRenameManager::AddRef()
+{
+    return InterlockedIncrement(&m_refCount);
+}
+
+IFACEMETHODIMP_(ULONG) CPowerRenameManager::Release()
+{
+    long refCount = InterlockedDecrement(&m_refCount);
+
+    if (refCount == 0)
+    {
+        delete this;
+    }
+    return refCount;
+}
+
+IFACEMETHODIMP CPowerRenameManager::QueryInterface(_In_ REFIID riid, _Outptr_ void** ppv)
+{
+    static const QITAB qit[] = {
+        QITABENT(CPowerRenameManager, IPowerRenameManager),
+        QITABENT(CPowerRenameManager, IPowerRenameRegExEvents),
+        { 0 }
+    };
+    return QISearch(this, qit, riid, ppv);
+}
+
+IFACEMETHODIMP CPowerRenameManager::Advise(_In_ IPowerRenameManagerEvents* renameOpEvents, _Out_ DWORD* cookie)
+{
+    CSRWExclusiveAutoLock lock(&m_lockEvents);
+    m_cookie++;
+    SMART_RENAME_MGR_EVENT srme;
+    srme.cookie = m_cookie;
+    srme.pEvents = renameOpEvents;
+    renameOpEvents->AddRef();
+    m_PowerRenameManagerEvents.push_back(srme);
+
+    *cookie = m_cookie;
+
+    return S_OK;
+}
+
+IFACEMETHODIMP CPowerRenameManager::UnAdvise(_In_ DWORD cookie)
+{
+    HRESULT hr = E_FAIL;
+    CSRWExclusiveAutoLock lock(&m_lockEvents);
+
+    for (std::vector<SMART_RENAME_MGR_EVENT>::iterator it = m_PowerRenameManagerEvents.begin(); it != m_PowerRenameManagerEvents.end(); ++it)
```

Comment:
Should be able to do:
for (auto it : m_PowerRenameManagerEvents)

Reply:
Thanks!  Done.